File: py_m/ast_.py
from abc import ABCMeta, abstractmethod
import token_


class Node(metaclass=ABCMeta):
    @abstractmethod
    def token_literal(self):
        pass

    @abstractmethod
    def string(self):
        pass


class Statement(Node):
    @abstractmethod
    def statement_node(self):
        pass


class Expression(Node):
    @abstractmethod
    def expression_node(self):
        pass


class Program(Node):
    def __init__(self):
        self.statements = []

    def token_literal(self):
        if len(self.statements) > 0:
            return self.statements[0].token_literal()
        else:
            return ""

    def string(self):
        s = ""
        for v in self.statements:
            s += v.string()
        return s

    def __str__(self):
        return "Program(Node)"


class LetStatement(Statement):
    """let文"""

    def __init__(self, token=None, name=None, value=None):
        self.token = token
        self.name = name
        self.value = value

    def token_literal(self):
        return self.token.literal

    def statement_node(self):
        pass

    def string(self):
        out = self.token_literal()
        out += " "
        out += self.name.string()
        out += " = "
        if self.value is not None:
            out += self.value.string()
        out += ";"
        return out

    def __str__(self):
        return "LetStatement(Statement)"


class Identifier(Expression):
    """識別子"""

    def __init__(self, token=None, value=None):
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        return self.value

    def __str__(self):
        return "Identifier(Expression)"


class ReturnStatement(Statement):
    """return文"""

    def __init__(self, token=None, return_value=None):
        self.token = token
        self.return_value = return_value

    def token_literal(self):
        return self.token.literal

    def statement_node(self):
        pass

    def string(self):
        out = self.token_literal()
        out += " "
        if self.return_value is not None:
            out += self.return_value.string()
        out += ";"
        return out

    def __str__(self):
        return "ReturnStatement(Statement)"


class ExpressionStatement(Statement):
    """式文"""

    def __init__(self, token=None, expression=None):
        self.token = token
        self.expression = expression

    def token_literal(self):
        return self.token.literal

    def statement_node(self):
        pass

    def string(self):
        if self.expression is not None:
            return self.expression.string()

        return ""

    def __str__(self):
        return "ExpressionStatement(Statement)"


class IntegerLiteral(Expression):
    """整数"""

    def __init__(self, token=None, value=None):
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        return self.token.literal

    def __str__(self):
        return "IntegerLiteral(Expression)"


class FloatLiteral(Expression):
    """実数"""

    def __init__(self, token=None, value=None):
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        return self.token.literal

    def __str__(self):
        return "FloatLiteral(Expression)"


class PrefixExpression(Expression):
    """前置演算子"""

    def __init__(self, token=None, operator="", right=None):
        self.token = token
        # "-", "!" が来る
        self.operator = operator
        self.right = right

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        out = "("
        out += self.operator
        out += self.right.string()
        out += ")"
        return out

    def __str__(self):
        return "PrefixExpression(Expression)"


class InfixExpression(Expression):
    """中置演算子"""

    def __init__(self, token=None, left=None, operator="", right=None):
        self.token = token
        self.left = left
        self.operator = operator
        self.right = right

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        out = "("
        out += self.left.string()
        out += " "
        out += self.operator
        out += " "
        out += self.right.string()
        out += ")"
        return out

    def __str__(self):
        return "InfixExpression(Expression)"


class Boolean(Expression):
    """真偽値"""

    def __init__(self, token=None, value=False):
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        return self.token.literal

    def __str__(self):
        return "Boolean(Expression)"


class IfExpression(Expression):
    """if式"""

    def __init__(self, token=None, condition=None, consequence=None, alternative=None):
        self.token = token
        self.condition = condition
        self.consequence = consequence
        self.alternative = alternative

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        out = "if"
        out += self.condition.string()
        out += " "
        out += self.consequence.string()
        if self.alternative is not None:
            out += "else "
            out += self.alternative.string()
        return out

    def __str__(self):
        return "IfExpression(Expression)"


class BlockStatement(Statement):
    """ブロック文"""

    def __init__(self, token=None):
        self.token = token
        self.statements = []

    def token_literal(self):
        return self.token.literal

    def statement_node(self):
        pass

    def string(self):
        s = ""
        for v in self.statements:
            s += v.string()
        return s

    def __str__(self):
        return "BlockStatement(Statement)"


class FunctionLiteral(Expression):
    """関数リテラル"""

    def __init__(self, token=None, parameters=[], body=None):
        self.token = token
        # Identifier のリスト
        self.parameters = []
        # BlockStatement
        self.body = body

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        params = []
        for v in self.parameters:
            params.append(v.string())

        out = self.token_literal()
        out += "("
        out += ", ".join(params)
        out += ")"
        out += self.body.string()
        return out

    def __str__(self):
        return "FunctionLiteral(Expression)"


class CallExpression(Expression):
    """呼び出し式"""

    def __init__(self, token=None, function=None, arguments=[]):
        self.token = token
        # Identifier or FunctionLiteral
        self.function = function
        # []Expression
        self.arguments = arguments

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        args = []
        for v in self.arguments:
            args.append(v.string())

        out = self.function.string()
        out += "("
        out += ", ".join(args)
        out += ")"

        return out

    def __str__(self):
        return "CallExpression(Expression)"


class StringLiteral(Expression):
    """文字列"""

    def __init__(self, token=None, value=None):
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        return self.token.literal

    def __str__(self):
        return "StringLiteral(Expression)"


class ArrayLiteral(Expression):
    """配列リテラル"""

    def __init__(self, token=None, elements=[]):
        # the '[' token
        self.token: token_.TokenType = token
        # []Expression
        self.elements: list[Expression] = elements

    def token_literal(self):
        return self.token.literal

    def expression_node(self):
        pass

    def string(self):
        el = []
        for v in self.elements:
            el.append(v.string())

        out = "["
        out += ", ".join(el)
        out += "]"

        return out

    def __str__(self):
        return "ArrayLiteral(Expression)"


if __name__ == "__main__":
    from token_ import Token, TokenType
    ls = LetStatement(
        token=Token(TokenType.LET, "let"),
        name=Identifier(Token(TokenType.IDENT, "myVar"), "myVar"),
        value=Identifier(Token(TokenType.IDENT, "anothorVal"), "anothorVal")
    )
    p = Program()
    p.statements.append(ls)
    print(p.string())

File: nim_m/ast.nim
import strutils
import token


type
  Node* = ref object of RootObj
  Statement* = ref object of Node
  Expression* = ref object of Node

method tokenLiteral(self: Node): string{.base.} =
  result = "tokenLiteral(self: Node)"

method toString(self: Node): string{.base.} =
  result = "toString(self: Node)"


type Program* = ref object of Node
  # let文、return文、式文
  statements*: seq[Statement]

method tokenLiteral*(self: Program): string =
  if len(self.statements) > 0:
    return self.statements[0].tokenLiteral()
  ""

method toString*(self: Program): string =
  result = ""
  for v in self.statements:
    result.add(v.toString())


type Identifier* = ref object of Expression
  token*: Token
  value*: string

method tokenLiteral*(self: Identifier): string =
  self.token.literal

method toString*(self: Identifier): string =
  self.value


type LetStatement* = ref object of Statement
  token*: Token
  name*: Identifier
  value*: Expression

method tokenLiteral*(self: LetStatement): string =
  self.token.literal

method toString*(self: LetStatement): string =
  result = self.tokenLiteral() & " "
  result.add(self.name.toString())
  result.add(" = ")
  if self.value != nil:
    result.add(self.value.toString())
  result.add(";")


type ReturnStatement* = ref object of Statement
  token*: Token
  returnValue*: Expression

method tokenLiteral*(self: ReturnStatement): string =
  self.token.literal

method toString*(self: ReturnStatement): string =
  result = self.tokenLiteral() & " "
  if self.returnValue != nil:
    result.add(self.returnValue.toString())
  result.add(";")


type ExpressionStatement* = ref object of Statement
  token*: Token
  expression*: Expression

method tokenLiteral*(self: ExpressionStatement): string =
  self.token.literal

method toString*(self: ExpressionStatement): string =
  if self.expression != nil:
    return self.expression.toString()
  ""


type IntegerLiteral* = ref object of Expression
  token*: Token
  value*: int

method tokenLiteral*(self: IntegerLiteral): string =
  self.token.literal

method toString*(self: IntegerLiteral): string =
  self.token.literal


type FloatLiteral* = ref object of Expression
  token*: Token
  value*: float

method tokenLiteral*(self: FloatLiteral): string =
  self.token.literal

method toString*(self: FloatLiteral): string =
  self.token.literal


type PrefixExpression* = ref object of Expression
  token*: Token
  operator*: string
  right*: Expression

method tokenLiteral*(self: PrefixExpression): string =
  self.token.literal

method toString*(self: PrefixExpression): string =
  result = "("
  result.add(self.operator)
  result.add(self.right.toString())
  result.add(")")


type InfixExpression* = ref object of Expression
  token*: Token
  left*: Expression
  operator*: string
  right*: Expression

method tokenLiteral*(self: InfixExpression): string =
  self.token.literal

method toString*(self: InfixExpression): string =
  result = "("
  result.add(self.left.toString())
  result.add(" " & self.operator & " ")
  result.add(self.right.toString())
  result.add(")")


type Boolean* = ref object of Expression
  token*: Token
  value*: bool

method tokenLiteral*(self: Boolean): string =
  self.token.literal

method toString*(self: Boolean): string =
  self.token.literal


type BlockStatement* = ref object of Statement
  token*: Token
  statements*: seq[Statement]

method tokenLiteral*(self: BlockStatement): string =
  self.token.literal

method toString*(self: BlockStatement): string =
  result = ""
  for v in self.statements:
    result.add(v.toString())


type IfExpression* = ref object of Expression
  token*: Token
  condition*: Expression
  consequence*: BlockStatement
  alternative*: BlockStatement

method tokenLiteral*(self: IfExpression): string =
  self.token.literal

method toString*(self: IfExpression): string =
  result = "if"
  result.add(self.condition.toString())
  result.add(" ")
  result.add(self.consequence.toString())
  if self.alternative != nil:
    result.add("else ")
    result.add(self.alternative.toString())


type FunctionLiteral* = ref object of Expression
  token*: Token
  parameters*: seq[Identifier]
  body*: BlockStatement

method tokenLiteral*(self: FunctionLiteral): string =
  self.token.literal

method toString*(self: FunctionLiteral): string =
  var params = newSeq[string]()
  for v in self.parameters:
    params.add(v.toString())

  result = self.tokenLiteral()
  result.add("(")
  result.add(params.join(", "))
  result.add(")")
  result.add(self.body.toString())


type CallExpression* = ref object of Expression
  token*: Token
  function*: Expression
  arguments*: seq[Expression]

method tokenLiteral*(self: CallExpression): string =
  self.token.literal

method toString*(self: CallExpression): string =
  var args = newSeq[string]()
  for v in self.arguments:
    args.add(v.toString())

  result = self.function.toString()
  result.add("(")
  result.add(args.join(", "))
  result.add(")")


type StringLiteral* = ref object of Expression
  token*: Token
  value*: string

method tokenLiteral*(self: StringLiteral): string =
  self.token.literal

method toString*(self: StringLiteral): string =
  self.token.literal


type ArrayLiteral* = ref object of Expression
  # the '[' token
  token*: Token
  elements*: seq[Expression]

method tokenLiteral*(self: ArrayLiteral): string =
  self.token.literal

method toString*(self: ArrayLiteral): string =
  var el = newSeq[string]()
  for v in self.elements:
    el.add(v.toString())

  result = "["
  result.add(el.join(", "))
  result.add("]")





when isMainModule:
  var t = Token(tokenType: LET, literal: "let")
  # echo t[]
  var i = Identifier(token: t, value: "x")
  # echo i[]
  # echo i.tokenLiteral()
  # echo i.toString()

  var l = LetStatement(token: t, name: i, value: i)
  echo l[]
  echo l.token[]
  echo l.name[]
  echo l.value[]
  echo l.toString()
  echo 0
